_blogposts/compiler/overview-of-new_encoding.mdx

```diff
@@ -0,0 +1,245 @@
+---
+author: hongbo
+date: "2020-06-10 18:00" 
+previewImg:
+category: compiler
+title: Overview of new encoding in BuckleScript
+description: |
+  Highlights of our newest changes to the internal representation 
+  and how they will benefit our users.
+---
+
+
+---
+title: An overview of new data representation in BuckleScript
+---
+
+In the next version of BuckleScript, we will make several major changes to tweak the data representation for various data types, 
+making them more idiomatic and debugger friendly. 
+
+Note: since V8 or other JavaScript engines are tweaked to make idiomatic JS code run fast, these changes also results in faster running code. 
+
+Another property, for a compiled language like BuckleScript, 
+is that we can reason about [IC](https://en.wikipedia.org/wiki/Inline_caching) friendliness by just looking at the type definitions locally; 
+this is very helpful for advanced users to write performance predictable JS code. 
+If you are unfamiliar with how IC works in general, feel free to skip the section about IC.
```

Comment:
I kinda feel like there should be a very short introduction on what IC actually means in a quick section... otherwise ppl will either 1) jump to wikipedia and read that IC thing instead of reading the article, or 2) read the article without the whole context of how IC relates to the compilation output.

It doesn't have to be scientifically correct, but just giving one simple example on how same shaped structures help the JS runtime to do performance optimizations would already be enough to make this article easier to understand.

Example:

## Why is Inline Caching Relevant?

JS engines are really clever in guessing the shapes of the data they are handling. E.g. when you are calling a function with a certain argument, such as:

```js
getName({name: "test1"});
getName({name: "test2"});
```

For each function call of `getName`, the JS engine gets more insights on what data structures the function is handling and keeps track of their shapes in the inline cache. So as long as we use singly-shaped ("monomorphic") values whenever we call a function, execution will be generally faster since the runtime doesn't need to look up and verify the shape all the time.

That's why BuckleScript will try to maintain the same shape for object data types whenever possible. For instance, when you are using a record:

```
type user = {
  name: string,
  age: option(int),
};

let myUser = { name: "BuckleScript", age: None};

let getName = (u) => u.name;

getName(myUser);
```

it will maintain always keep all the attributes of `myUser`, even though `age` is `None`:

```
var myUser = {
  name: "BuckleScript",
  age: undefined
};

getName(myUser);
```

This behavior will make it very easy for Reason developers to reliably build data structures for maximum runtime performance.

This classification algorithm goes for all kinds of data structures such as arrays as well. If you are maintaining an array such as `["1", "2", "3"]`, JS runtimes like V8 will classify it as a so-called `monomorphic` data type, since it only contains strings. If you are creating an array `["1", 2, {num: "3"}]` it will be `polymorphic` and will be harder to optimize during runtime. (Please note that this is an extremely simplified explanation just to get the idea across).

Note that Reason is a monomorphic language, so we are generally forced to build data structures that can only be in one shape (e.g. variants or `array(string)`).

Reply:
There're subtle differences here, the call-site monomorphism is subtly different from what we used to talk about, if we expand on this, it will goes to a couple of blog posts

Reply:
Yeah, I am sure this topic is indefinitely complex, just thought it would be great to get kickstarted with some really condensed low-scientific information to get the point across and then link to the extensive Wiki article etc for further read (for those who like the hardcore stuff).

Feel free to ignore this

Reply:
added a tutorial link